# P03 - Final
- Dayanni Godoy Rosales

- Mateo Tavares Trueba

- Jesús Flores Cortes

## Objetivo General

Predecir con claridad la popularidad de la publicación con un dataset que contiene caracteríisticas de notas periodisticas y su respectivo alcance en redes sociales.

Evaluar el desempeño predictivo de modelos avanzados de aprendizaje automático —Random Forest y XGBoost, tanto en clasificación como en regresión— utilizando validación cruzada estratificada (Stratified K-Folds), con el fin de identificar la capacidad real de las variables disponibles para explicar y predecir el fenómeno de interés.

## Objetivos Específicos

- Aplicar la técnica de Stratified K-Fold Cross-Validation para garantizar una evaluación robusta, estable y representativa del desempeño de los modelos, preservando la proporción de clases en cada partición del conjunto de datos.

- Entrenar y evaluar un modelo Random Forest de regresión, con el propósito de estimar la capacidad explicativa de las características y determinar hasta qué punto los patrones no lineales permiten predecir la variable objetivo continua.

- Entrenar y evaluar un modelo XGBoost de regresión, optimizando sus hiperparámetros para comparar su desempeño frente al Random Forest y analizar si técnicas de boosting mejoran la precisión en escenarios de alta complejidad o ruido.

- Entrenar y evaluar un modelo Random Forest de clasificación, con el objetivo de identificar la habilidad del algoritmo para discriminar entre clases utilizando métricas robustas como ROC AUC y accuracy.

- Entrenar y evaluar un modelo XGBoost de clasificación, buscando determinar si los métodos de boosting logran un mejor rendimiento que los métodos de bagging en la identificación de patrones no lineales y relaciones complejas entre las variables.

- Comparar de manera sistemática los resultados de regresión y clasificación, con el fin de determinar cuál enfoque predictivo se ajusta mejor al comportamiento real de los datos y si la naturaleza del fenómeno favorece estimaciones continuas o categóricas.

- Analizar la estabilidad de los modelos mediante la revisión de la desviación estándar de las métricas obtenidas durante la validación cruzada, para evaluar la generalización y riesgo de sobreajuste.

In [76]:
import pandas as pd
import numpy as np

In [77]:
df = pd.read_csv('OnlineNewsPopularity.csv')

In [78]:
df.head()

url   timedelta   n_tokens_title   n_tokens_content   n_unique_tokens   n_non_stop_words   n_non_stop_unique_tokens   num_hrefs   num_self_hrefs   num_imgs   num_videos   average_token_length   num_keywords   data_channel_is_lifestyle   data_channel_is_entertainment   data_channel_is_bus   data_channel_is_socmed   data_channel_is_tech   data_channel_is_world   kw_min_min   kw_max_min   kw_avg_min   kw_min_max   kw_max_max   kw_avg_max   kw_min_avg   kw_max_avg   kw_avg_avg   self_reference_min_shares   self_reference_max_shares   self_reference_avg_sharess   weekday_is_monday   weekday_is_tuesday   weekday_is_wednesday   weekday_is_thursday   weekday_is_friday   weekday_is_saturday   weekday_is_sunday   is_weekend    LDA_00    LDA_01    LDA_02    LDA_03    LDA_04   global_subjectivity   global_sentiment_polarity   global_rate_positive_words   global_rate_negative_words   rate_positive_words   rate_negative_words  \
0  http://mashable.com/2013/01/07/amazon-instant-...       731.0             12.0              219.0          0.663594                1.0                   0.815385         4.0              2.0        1.0          0.0               4.680365            5.0                         0.0                             1.0                   0.0                      0.0                    0.0                     0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0                       496.0                       496.0                   496.000000                 1.0                  0.0                    0.0                   0.0                 0.0                   0.0                 0.0          0.0  0.500331  0.378279  0.040005  0.041263  0.040123              0.521617                    0.092562                     0.045662                     0.013699              0.769231              0.230769   
1  http://mashable.com/2013/01/07/ap-samsung-spon...       731.0              9.0              255.0          0.604743                1.0                   0.791946         3.0              1.0        1.0          0.0               4.913725            4.0                         0.0                             0.0                   1.0                      0.0                    0.0                     0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0                         0.0                         0.0                     0.000000                 1.0                  0.0                    0.0                   0.0                 0.0                   0.0                 0.0          0.0  0.799756  0.050047  0.050096  0.050101  0.050001              0.341246                    0.148948                     0.043137                     0.015686              0.733333              0.266667   
2  http://mashable.com/2013/01/07/apple-40-billio...       731.0              9.0              211.0          0.575130                1.0                   0.663866         3.0              1.0        1.0          0.0               4.393365            6.0                         0.0                             0.0                   1.0                      0.0                    0.0                     0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0          0.0                       918.0                       918.0                   918.000000                 1.0                  0.0                    0.0                   0.0                 0.0                   0.0                 0.0          0.0  0.217792  0.033334  0.033351  0.033334  0.682188              0.702222                    0.323333                     0.056872                     0.009479              0.857143              0.142857   
3  http://mashable.com/2013/01/07/astronaut-notre...       731.0              9.0              531.0          0.503788                1.0

In [79]:
df.columns = df.columns.str.strip()

In [80]:
df.head()

url  timedelta  n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  num_videos  average_token_length  num_keywords  data_channel_is_lifestyle  data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  data_channel_is_tech  data_channel_is_world  kw_min_min  kw_max_min  kw_avg_min  kw_min_max  kw_max_max  kw_avg_max  kw_min_avg  kw_max_avg  kw_avg_avg  self_reference_min_shares  self_reference_max_shares  self_reference_avg_sharess  weekday_is_monday  weekday_is_tuesday  weekday_is_wednesday  weekday_is_thursday  weekday_is_friday  weekday_is_saturday  weekday_is_sunday  is_weekend    LDA_00    LDA_01    LDA_02    LDA_03    LDA_04  global_subjectivity  global_sentiment_polarity  global_rate_positive_words  global_rate_negative_words  rate_positive_words  rate_negative_words  avg_positive_polarity  min_positive_polarity  \
0  http://mashable.com/2013/01/07/amazon-instant-...      731.0            12.0             219.0         0.663594               1.0                  0.815385        4.0             2.0       1.0         0.0              4.680365           5.0                        0.0                            1.0                  0.0                     0.0                   0.0                    0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0                      496.0                      496.0                  496.000000                1.0                 0.0                   0.0                  0.0                0.0                  0.0                0.0         0.0  0.500331  0.378279  0.040005  0.041263  0.040123             0.521617                   0.092562                    0.045662                    0.013699             0.769231             0.230769               0.378636               0.100000   
1  http://mashable.com/2013/01/07/ap-samsung-spon...      731.0             9.0             255.0         0.604743               1.0                  0.791946        3.0             1.0       1.0         0.0              4.913725           4.0                        0.0                            0.0                  1.0                     0.0                   0.0                    0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0                        0.0                        0.0                    0.000000                1.0                 0.0                   0.0                  0.0                0.0                  0.0                0.0         0.0  0.799756  0.050047  0.050096  0.050101  0.050001             0.341246                   0.148948                    0.043137                    0.015686             0.733333             0.266667               0.286915               0.033333   
2  http://mashable.com/2013/01/07/apple-40-billio...      731.0             9.0             211.0         0.575130               1.0                  0.663866        3.0             1.0       1.0         0.0              4.393365           6.0                        0.0                            0.0                  1.0                     0.0                   0.0                    0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0                      918.0                      918.0                  918.000000                1.0                 0.0                   0.0                  0.0                0.0                  0.0                0.0         0.0  0.217792  0.033334  0.033351  0.033334  0.682188             0.702222                   0.323333                    0.056872                    0.009479             0.857143             0.142857               0.495833               0.100000   
3  http://mashable.com/2013/01/07/astronaut-notre...      731.0             9.0             531.0         0.503788               

In [81]:
cols_to_drop = ['url', 'timedelta']
df_clean = df.drop(columns=cols_to_drop)

In [82]:
cols_binarias = [col for col in df_clean.columns if 'weekday_is' in col or 'data_channel_is' in col]

for col in cols_binarias:
    df_clean[col] = df_clean[col].astype(int)

In [83]:
median_shares = df_clean['shares'].median()

In [84]:
df_clean['is_popular'] = (df_clean['shares'] >= median_shares).astype(int)


In [85]:
df_clean['log_shares'] = np.log1p(df_clean['shares'])

In [86]:
df_clean.describe()

n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  n_non_stop_unique_tokens     num_hrefs  num_self_hrefs      num_imgs    num_videos  average_token_length  num_keywords  data_channel_is_lifestyle  data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  data_channel_is_tech  data_channel_is_world    kw_min_min     kw_max_min    kw_avg_min     kw_min_max     kw_max_max     kw_avg_max    kw_min_avg     kw_max_avg    kw_avg_avg  self_reference_min_shares  self_reference_max_shares  self_reference_avg_sharess  weekday_is_monday  weekday_is_tuesday  weekday_is_wednesday  weekday_is_thursday  weekday_is_friday  weekday_is_saturday  weekday_is_sunday    is_weekend        LDA_00        LDA_01        LDA_02        LDA_03        LDA_04  global_subjectivity  global_sentiment_polarity  global_rate_positive_words  global_rate_negative_words  rate_positive_words  rate_negative_words  avg_positive_polarity  min_positive_polarity  max_positive_polarity  \
count    39644.000000      39644.000000     39644.000000      39644.000000              39644.000000  39644.000000    39644.000000  39644.000000  39644.000000          39644.000000  39644.000000               39644.000000                   39644.000000         39644.000000            39644.000000          39644.000000           39644.000000  39644.000000   39644.000000  39644.000000   39644.000000   39644.000000   39644.000000  39644.000000   39644.000000  39644.000000               39644.000000               39644.000000                39644.000000       39644.000000        39644.000000          39644.000000         39644.000000       39644.000000         39644.000000       39644.000000  39644.000000  39644.000000  39644.000000  39644.000000  39644.000000  39644.000000         39644.000000               39644.000000                39644.000000                39644.000000         39644.000000         39644.000000           39644.000000           39644.000000           39644.000000   
mean        10.398749        546.514731         0.548216          0.996469                  0.689175     10.883690        3.293638      4.544143      1.249874              4.548239      7.223767                   0.052946                       0.178009             0.157855                0.058597              0.185299               0.212567     26.106801    1153.951682    312.366967   13612.354102  752324.066694  259281.938083   1117.146610    5657.211151   3135.858639                3998.755396               10329.212662                 6401.697580           0.168020            0.186409              0.187544             0.183306           0.143805             0.061876           0.069039      0.130915      0.184599      0.141256      0.216321      0.223770      0.234029             0.443370                   0.119309                    0.039625                    0.016612             0.682150             0.287934               0.353825               0.095446               0.756728   
std          2.114037        471.107508         3.520708          5.231231                  3.264816     11.332017        3.855141      8.309434      4.107855              0.844406      1.909130                   0.223929                       0.382525             0.364610                0.234871              0.388545               0.409129     69.633215    3857.990877    620.783887   57986.029357  214502.129573  135102.247285   1137.456951    6098.871957   1318.150397               19738.670516               41027.576613                24211.332231           0.373889            0.389441              0.390353             0.386922           0.350896             0.240933           0.253524      0.337312      0.262975      0.219707      0.282145      0.295191      0.289183             0.116685                   0.096931                    0.017429                    0.010828             0.190206             0.156156               0.104542               0.071315               0.247786   
min        

In [87]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39644 entries, 0 to 39643
Data columns (total 61 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   n_tokens_title                 39644 non-null  float64
 1   n_tokens_content               39644 non-null  float64
 2   n_unique_tokens                39644 non-null  float64
 3   n_non_stop_words               39644 non-null  float64
 4   n_non_stop_unique_tokens       39644 non-null  float64
 5   num_hrefs                      39644 non-null  float64
 6   num_self_hrefs                 39644 non-null  float64
 7   num_imgs                       39644 non-null  float64
 8   num_videos                     39644 non-null  float64
 9   average_token_length           39644 non-null  float64
 10  num_keywords                   39644 non-null  float64
 11  data_channel_is_lifestyle      39644 non-null  int64  
 12  data_channel_is_entertainment  39644 non-null 

In [88]:
cols_to_exclude = ['shares', 'log_shares', 'is_popular']

In [89]:
X = df_clean.drop(columns=cols_to_exclude)

In [90]:
y_reg = df_clean['log_shares']      #esta es para regresion mateo
y_class = df_clean['is_popular']    #esta es para clasificacion 

In [91]:
import pandas as pd


pd.set_option('display.width', 1000) 

print(df_clean.head())

   n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  num_videos  average_token_length  num_keywords  data_channel_is_lifestyle  data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  data_channel_is_tech  data_channel_is_world  kw_min_min  kw_max_min  kw_avg_min  kw_min_max  kw_max_max  kw_avg_max  kw_min_avg  kw_max_avg  kw_avg_avg  self_reference_min_shares  self_reference_max_shares  self_reference_avg_sharess  weekday_is_monday  weekday_is_tuesday  weekday_is_wednesday  weekday_is_thursday  weekday_is_friday  weekday_is_saturday  weekday_is_sunday  is_weekend    LDA_00    LDA_01    LDA_02    LDA_03    LDA_04  global_subjectivity  global_sentiment_polarity  global_rate_positive_words  global_rate_negative_words  rate_positive_words  rate_negative_words  avg_positive_polarity  min_positive_polarity  max_positive_polarity  avg_negative_polarity  min_negative_polarity  \
0            

# Modelos Propuestos. (Se integra el preprocesamiento al pipeline directamente)

## Clasificacion: Variable a clasificar 'Is Popular' 

Clasificaremos sin tener la información de los shares, es decir, a partir de toda la información sin tener los shares veremos si se hace popular o no.

In [92]:
df_clas= df_clean.drop(columns=["shares", "log_shares"])

In [93]:
df_clas.head()

n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  num_videos  average_token_length  num_keywords  data_channel_is_lifestyle  data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  data_channel_is_tech  data_channel_is_world  kw_min_min  kw_max_min  kw_avg_min  kw_min_max  kw_max_max  kw_avg_max  kw_min_avg  kw_max_avg  kw_avg_avg  self_reference_min_shares  self_reference_max_shares  self_reference_avg_sharess  weekday_is_monday  weekday_is_tuesday  weekday_is_wednesday  weekday_is_thursday  weekday_is_friday  weekday_is_saturday  weekday_is_sunday  is_weekend    LDA_00    LDA_01    LDA_02    LDA_03    LDA_04  global_subjectivity  global_sentiment_polarity  global_rate_positive_words  global_rate_negative_words  rate_positive_words  rate_negative_words  avg_positive_polarity  min_positive_polarity  max_positive_polarity  avg_negative_polarity  min_negative_polarity  \
0            12.0             219.0         0.663594               1.0                  0.815385        4.0             2.0       1.0         0.0              4.680365           5.0                          0                              1                    0                       0                     0                      0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0                      496.0                      496.0                  496.000000                  1                   0                     0                    0                  0                    0                  0         0.0  0.500331  0.378279  0.040005  0.041263  0.040123             0.521617                   0.092562                    0.045662                    0.013699             0.769231             0.230769               0.378636               0.100000                    0.7              -0.350000                 -0.600   
1             9.0             255.0         0.604743               1.0                  0.791946        3.0             1.0       1.0         0.0              4.913725           4.0                          0                              0                    1                       0                     0                      0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0                        0.0                        0.0                    0.000000                  1                   0                     0                    0                  0                    0                  0         0.0  0.799756  0.050047  0.050096  0.050101  0.050001             0.341246                   0.148948                    0.043137                    0.015686             0.733333             0.266667               0.286915               0.033333                    0.7              -0.118750                 -0.125   
2             9.0             211.0         0.575130               1.0                  0.663866        3.0             1.0       1.0         0.0              4.393365           6.0                          0                              0                    1                       0                     0                      0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0         0.0                      918.0                      918.0                  918.000000                  1                   0                     0                    0                  0                    0                  0         0.0  0.217792  0.033334  0.033351  0.033334  0.682188             0.702222                   0.323333                    0.056872                    0.009479             0.857143             0.142857               0.495833               0.100000                    1.0              -0.466667                 -0.800   
3             9.0             531.0         0.503788   

In [94]:
import pandas as pd
import optuna
from xgboost import XGBClassifier
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.model_selection import cross_validate, StratifiedKFold


target_column = 'is_popular'

X = df_clas.drop(columns=[target_column])
y = df_clas[target_column]

num_cols = X.select_dtypes(include=['number']).columns.tolist()

numerical_transformer = StandardScaler()


preprocessor = ColumnTransformer(transformers=[
    ('num', numerical_transformer, num_cols)     
])


def objective(trial):
    
    param_grid = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 800),
        'learning_rate': trial.suggest_float('learning_rate', 0.005, 0.3, log=True),
        'max_depth': trial.suggest_int('max_depth', 3, 10),
        'subsample': trial.suggest_float('subsample', 0.6, 1.0),
    }
    
    model = XGBClassifier(
        **param_grid, 
        random_state=42, 
        n_jobs=-1, 
        eval_metric='logloss'
    )
    
    pipeline = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('model', model)
    ])
    

    cv_inner = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
    
   
    cv_results = cross_validate(
        pipeline, 
        X, 
        y, 
        cv=cv_inner, 
        scoring=['accuracy', 'roc_auc'],
        n_jobs=-1
    )
    
    mean_auc = cv_results['test_roc_auc'].mean()
    mean_accuracy = cv_results['test_accuracy'].mean()
    
    std_auc = cv_results['test_roc_auc'].std()
    std_accuracy = cv_results['test_accuracy'].std()

    trial.set_user_attr("accuracy", mean_accuracy)
    trial.set_user_attr("std_auc", std_auc)       
    trial.set_user_attr("std_accuracy", std_accuracy)
    
    return mean_auc


print("Iniciando optimización...")
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=30) 

print(f"Mejor ROC AUC: {study.best_value:.4f}")
print(f"Desviación Estándar ROC AUC: {study.best_trial.user_attrs['std_auc']:.4f}")

print(f"Accuracy del mejor modelo: {study.best_trial.user_attrs['accuracy']:.4f}")
print(f"Desviación Estándar Accuracy: {study.best_trial.user_attrs['std_accuracy']:.4f}")


[I 2025-12-05 16:36:35,363] A new study created in memory with name: no-name-a14fc568-9ca6-4290-a977-6339448a3ee6


Iniciando optimización...


[I 2025-12-05 16:36:56,748] Trial 0 finished with value: 0.7355476958636649 and parameters: {'n_estimators': 746, 'learning_rate': 0.033862500461139096, 'max_depth': 7, 'subsample': 0.7002754093534714}. Best is trial 0 with value: 0.7355476958636649.
[I 2025-12-05 16:37:14,169] Trial 1 finished with value: 0.7220883845891123 and parameters: {'n_estimators': 452, 'learning_rate': 0.17200469488292972, 'max_depth': 9, 'subsample': 0.9909626630820184}. Best is trial 0 with value: 0.7355476958636649.
[I 2025-12-05 16:37:20,650] Trial 2 finished with value: 0.7238427980686378 and parameters: {'n_estimators': 340, 'learning_rate': 0.007926813987947028, 'max_depth': 4, 'subsample': 0.8917498138230726}. Best is trial 0 with value: 0.7355476958636649.
[I 2025-12-05 16:37:38,490] Trial 3 finished with value: 0.7197095134255035 and parameters: {'n_estimators': 584, 'learning_rate': 0.11502516705628378, 'max_depth': 8, 'subsample': 0.7078023655477784}. Best is trial 0 with value: 0.7355476958636649

Mejor ROC AUC: 0.7409
Desviación Estándar ROC AUC: 0.0032
Accuracy del mejor modelo: 0.6782
Desviación Estándar Accuracy: 0.0040


In [95]:
print(f"Mejores parámetros: {study.best_params}")

Mejores parámetros: {'n_estimators': 487, 'learning_rate': 0.032852784434223965, 'max_depth': 5, 'subsample': 0.8289747542166606}



### Resultado XGBoost de clasificación:
- Mejor ROC AUC: 0.7409
- Desviación Estándar ROC AUC: 0.0032
- Mejores parámetros: 'n_estimators': 487, 'learning_rate': 0.032852784434223965, 'max_depth': 5, 'subsample': 0.8289747542166606
- Accuracy del mejor modelo:  0.6782
- Desviación Estándar Accuracy: 0.0040



In [96]:
import pandas as pd
import optuna
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.model_selection import cross_validate, StratifiedKFold

target_column = 'is_popular'

X = df_clas.drop(columns=[target_column])
y = df_clas[target_column]

num_cols = X.select_dtypes(include=['number']).columns.tolist()

numerical_transformer = StandardScaler()


preprocessor = ColumnTransformer(transformers=[
    ('num', numerical_transformer, num_cols)     
])
def objective(trial):
    
    param_grid = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 800),
        'max_depth': trial.suggest_int('max_depth', 5, 50),
        'min_samples_split': trial.suggest_int('min_samples_split', 2, 20),
        'min_samples_leaf': trial.suggest_int('min_samples_leaf', 1, 10),
    }
    
    model = RandomForestClassifier(
        **param_grid, 
        random_state=42, 
        n_jobs=-1
    )
    
    pipeline = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('model', model)
    ])
    
    cv_inner = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
    
    cv_results = cross_validate(
        pipeline, 
        X, 
        y, 
        cv=cv_inner, 
        scoring=['accuracy', 'roc_auc'],
        n_jobs=-1
    )
    
    mean_auc = cv_results['test_roc_auc'].mean()
    mean_accuracy = cv_results['test_accuracy'].mean()

    std_auc = cv_results['test_roc_auc'].std()
    std_accuracy = cv_results['test_accuracy'].std()

    trial.set_user_attr("accuracy", mean_accuracy)
    trial.set_user_attr("std_auc", std_auc)     
    trial.set_user_attr("std_accuracy", std_accuracy)
    
    return mean_auc

print("Iniciando optimización...")
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=30) 

print(f"Mejor ROC AUC: {study.best_value:.4f}")
print(f"Mejores parámetros: {study.best_params}")

print(f"Mejor ROC AUC: {study.best_value:.4f}")
print(f"Desviación Estándar ROC AUC: {study.best_trial.user_attrs['std_auc']:.4f}")

print(f"Accuracy del mejor modelo: {study.best_trial.user_attrs['accuracy']:.4f}")
print(f"Desviación Estándar Accuracy: {study.best_trial.user_attrs['std_accuracy']:.4f}")
print(f"Mejores parámetros: {study.best_params}")

[I 2025-12-05 16:42:18,866] A new study created in memory with name: no-name-744a508c-983e-4c77-bc89-1ae9184c6f6a


Iniciando optimización...


[I 2025-12-05 16:42:30,268] Trial 0 finished with value: 0.7322084874422748 and parameters: {'n_estimators': 241, 'max_depth': 14, 'min_samples_split': 7, 'min_samples_leaf': 3}. Best is trial 0 with value: 0.7322084874422748.
[I 2025-12-05 16:42:46,513] Trial 1 finished with value: 0.7215758927906059 and parameters: {'n_estimators': 630, 'max_depth': 7, 'min_samples_split': 11, 'min_samples_leaf': 3}. Best is trial 0 with value: 0.7322084874422748.
[I 2025-12-05 16:43:02,121] Trial 2 finished with value: 0.7317790949750511 and parameters: {'n_estimators': 271, 'max_depth': 31, 'min_samples_split': 6, 'min_samples_leaf': 2}. Best is trial 0 with value: 0.7322084874422748.
[I 2025-12-05 16:43:20,582] Trial 3 finished with value: 0.7270553873158068 and parameters: {'n_estimators': 572, 'max_depth': 9, 'min_samples_split': 7, 'min_samples_leaf': 10}. Best is trial 0 with value: 0.7322084874422748.
[I 2025-12-05 16:43:30,223] Trial 4 finished with value: 0.7302590618043616 and parameters: 

Mejor ROC AUC: 0.7342
Mejores parámetros: {'n_estimators': 518, 'max_depth': 40, 'min_samples_split': 20, 'min_samples_leaf': 4}
Mejor ROC AUC: 0.7342
Desviación Estándar ROC AUC: 0.0029
Accuracy del mejor modelo: 0.6718
Desviación Estándar Accuracy: 0.0029
Mejores parámetros: {'n_estimators': 518, 'max_depth': 40, 'min_samples_split': 20, 'min_samples_leaf': 4}



### Resultado Random Forest de clasificación:
- Mejor ROC AUC: 0.7346
- Desviación Estándar ROC AUC: 0.0030
- Mejores parámetros:   {'n_estimators': 794, 'max_depth': 43, 'min_samples_split': 7, 'min_samples_leaf': 3}
- Accuracy del mejor modelo: 0.6729
- Desviación Estándar Accuracy: 0.0036

## Por poco, pero el ganador es XGBoost con .7409 de ROC-AUC y .67 de accuracy.

## Regresion: Variable objetivo: Log-Shares.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [ ]:
sns.histplot(df_clean['shares'], bins=50)
plt.title("Distribución Original")
plt.show()

In [ ]:
df_reg= df_clean.drop(columns=["shares", "is_popular"])

In [ ]:
sns.histplot(df_reg['log_shares'], bins=50)
plt.title("Distribución Transformada")
plt.show()

Hicimos una transformación logarítmica log-shares para que sean más rápidos los modelos.


### Random Forest de Regresión.

In [ ]:
df_reg.head()

In [ ]:
import pandas as pd
import optuna
from sklearn.ensemble import RandomForestRegressor
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.model_selection import cross_validate, KFold

target_column = 'log_shares'

X = df_reg.drop(columns=[target_column])
y = df_reg[target_column]

num_cols = X.select_dtypes(include=['number']).columns.tolist()

numerical_transformer = StandardScaler()


preprocessor = ColumnTransformer(transformers=[
    ('num', numerical_transformer, num_cols)     
])

def objective(trial):
    
    param_grid = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 800),
        'max_depth': trial.suggest_int('max_depth', 5, 50),
        'min_samples_split': trial.suggest_int('min_samples_split', 2, 20),
        'min_samples_leaf': trial.suggest_int('min_samples_leaf', 1, 10),
    }

    model = RandomForestRegressor(
        **param_grid, 
        random_state=42, 
        n_jobs=-1
    )
    
    pipeline = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('model', model)
    ])
    
    cv_inner = KFold(n_splits=5, shuffle=True, random_state=42)

    cv_results = cross_validate(
        pipeline, 
        X, 
        y, 
        cv=cv_inner, 
        scoring=['r2'], 
        n_jobs=-1
    )
    
    
    mean_r2 = cv_results['test_r2'].mean()
    std_r2 = cv_results['test_r2'].std()

    trial.set_user_attr("std_r2", std_r2)
    
    return mean_r2

print("Iniciando optimización para Regresión...")
study = optuna.create_study(direction='maximize') 
study.optimize(objective, n_trials=30) 

print(f"Mejor R2 Score: {study.best_value:.4f}")
print(f"Mejores parámetros: {study.best_params}")
print(f"Desviación Estándar del R2: {study.best_trial.user_attrs['std_r2']:.4f}")

In [62]:

print(f"Mejor R2 Score conseguido: {study.best_value:.4f}")
print(f"Mejores parámetros: {study.best_params}")
try:
    print(f"Desviación Estándar: {study.best_trial.user_attrs['std_r2']:.4f}")
except:
    print("No se pudo recuperar la desviación estándar del mejor trial.")

Mejor R2 Score conseguido: 0.1647
Mejores parámetros: {'n_estimators': 717, 'max_depth': 24, 'min_samples_split': 12, 'min_samples_leaf': 8}
Desviación Estándar: 0.0085



### Resultado Random Forest de Regresión:
- Mejor R2 Score: 0.1647
- Desviación Estándar ROC AUC: 0.0085
- Mejores parámetros:   {'n_estimators': 717, 'max_depth': 24, 'min_samples_split': 12, 'min_samples_leaf': 8}


### XGboost de Regresión.

In [63]:
import pandas as pd
import optuna
from xgboost import XGBRegressor
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_validate, KFold

target_column = 'log_shares'

X = df_reg.drop(columns=[target_column])
y = df_reg[target_column]

num_cols = X.select_dtypes(include=['number']).columns.tolist()

numerical_transformer = StandardScaler()

preprocessor = ColumnTransformer(transformers=[
    ('num', numerical_transformer, num_cols)     
])

def objective(trial):
    
    param_grid = {
        'n_estimators': trial.suggest_int('n_estimators', 100, 1000),
        'max_depth': trial.suggest_int('max_depth', 3, 12),
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.3, log=True),
        'subsample': trial.suggest_float('subsample', 0.6, 1.0)
    }

    model = XGBRegressor(
        **param_grid, 
        random_state=42, 
        n_jobs=-1,
        objective='reg:squarederror'
    )
    
    pipeline = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('model', model)
    ])
    
    cv_inner = KFold(n_splits=5, shuffle=True, random_state=42)

    cv_results = cross_validate(
        pipeline, 
        X, 
        y, 
        cv=cv_inner, 
        scoring=['r2'], 
        n_jobs=-1
    )
    
    mean_r2 = cv_results['test_r2'].mean()
    std_r2 = cv_results['test_r2'].std()

    trial.set_user_attr("std_r2", std_r2)
    
    return mean_r2

print("Iniciando optimización para Regresión con XGBoost...")
study = optuna.create_study(direction='maximize') 
study.optimize(objective, n_trials=10) 

print(f"Mejor R2 Score: {study.best_value:.4f}")
print(f"Mejores parámetros: {study.best_params}")
print(f"Desviación Estándar del R2: {study.best_trial.user_attrs['std_r2']:.4f}")

[I 2025-12-05 14:51:44,996] A new study created in memory with name: no-name-5a01eb9b-61a1-4100-a33d-d64ae499d3db


Iniciando optimización para Regresión con XGBoost...


[I 2025-12-05 14:51:56,130] Trial 0 finished with value: 0.1738734000734372 and parameters: {'n_estimators': 633, 'max_depth': 4, 'learning_rate': 0.036328857202027984, 'subsample': 0.7486015740412126}. Best is trial 0 with value: 0.1738734000734372.
[I 2025-12-05 14:52:01,797] Trial 1 finished with value: 0.1662852396251166 and parameters: {'n_estimators': 584, 'max_depth': 4, 'learning_rate': 0.09408386008135389, 'subsample': 0.9516335290485666}. Best is trial 0 with value: 0.1738734000734372.
[I 2025-12-05 14:52:08,373] Trial 2 finished with value: 0.17309545593841064 and parameters: {'n_estimators': 946, 'max_depth': 3, 'learning_rate': 0.03931827615325188, 'subsample': 0.8235405185150035}. Best is trial 0 with value: 0.1738734000734372.
[I 2025-12-05 14:52:19,849] Trial 3 finished with value: 0.07270298519088655 and parameters: {'n_estimators': 355, 'max_depth': 8, 'learning_rate': 0.16737147543727185, 'subsample': 0.6751685866908684}. Best is trial 0 with value: 0.173873400073437

Mejor R2 Score: 0.1739
Mejores parámetros: {'n_estimators': 633, 'max_depth': 4, 'learning_rate': 0.036328857202027984, 'subsample': 0.7486015740412126}
Desviación Estándar del R2: 0.0081



### Resultado XGboost de Regresión:
- Mejor R2 Score: 0.1739
- Desviación Estándar ROC AUC: 0.0081
- Mejores parámetros:  {'n_estimators': 633, 'max_depth': 4, 'learning_rate': 0.036328857202027984, 'subsample': 0.7486015740412126}

### Conclusiones

El mejor resultado corresponde al modelo XGBoost de regresión, que alcanzó un R² de 0.1739, superando ligeramente a Random Forest. Este valor indica que menos del 18% de la variabilidad en el número de shares puede ser explicada por las características disponibles. Esto sugiere que, a pesar de la riqueza del dataset, existen factores externos al contenido textual, metadatos o rendimiento histórico de keywords que influyen significativamente en la difusión real de una noticia. La baja capacidad explicativa también es consistente con la alta dispersión del comportamiento viral, fenómeno que suele responder a dinámicas sociales impredecibles, efectos de red, temporalidad y exposición inicial elementos no capturados por el conjunto actual de predictores.

Por otra parte, los resultados de clasificación muestran un desempeño más satisfactorio. El modelo XGBoost de clasificación obtuvo el mejor ROC AUC: 0.7411 y un accuracy de 0.6781, superando a Random Forest. Un ROC AUC por encima de 0.74 indica una capacidad moderada para distinguir entre artículos virales y no virales, lo cual refleja que, aunque predecir el número exacto de shares es complejo, sí existen patrones identificables que permiten clasificar si un artículo superará la mediana de viralidad. Variables relacionadas con el tema, temporalidad y desempeño histórico de keywords parecen aportar información suficiente para esta tarea, aunque no para predecir la magnitud exacta del impacto.

Además, las desviaciones estándar bajas en ambos enfoques evidencian estabilidad en las estimaciones, lo que demuestra que el uso de Stratified K-Folds fue adecuado para obtener métricas consistentes y evitar dependencia de una sola partición del dataset.